### PR TITLE
Move cloud provider controllers to k8s.io/cloud-provider 

### LINKED ISSUE #81172: Move cloud controllers to k8s.io/cloud-provider/controllers
**What would you like to be added**:
More cloud providers are adopting/using the out-of-tree cloud providers. One of the various issues / complaints we get is that cmd/cloud-controller-manager is difficult to consume because of the way it pulls in a bunch of other dependencies. 

**Why is this needed**:
I would like to see `cmd/cloud-controller-manager` in staging so it's easier to consume for external providers, however, because it imports controllers in `pkg/controllers/*` it can't be moved there yet. If we also moved the cloud controllers to somewhere like `k8s.io/cloud-provider/controllers`, that would unblock us from staging cmd/cloud-controller-manager as well.

### PR DESCRIPTION
**Pull request by andrewsykim**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
We're trying to make `cmd/cloud-controller-manager` easier to consume (https://github.com/kubernetes/cloud-provider/issues/29). This PR moves all the cloud provider specific controllers to k8s.io/cloud-provider as part of moving all of `cmd/cloud-controller-manager` eventually.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubernetes/issues/81172

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go [hunk 1]
```diff
@@ -1392,6 +1347,13 @@ func TestPatchStatus(t *testing.T) {
 }
 
 func Test_getNodeConditionPredicate(t *testing.T) {
+	// init feature gates, feature spec does not matter since it will be overriden by tests
```
Comment (cici37) [inline staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:1350]: Misspelling here of "overridden". And I think you mentioned that the test changes are moved to another pr #90987

Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:1350]: Good catch, will update after https://github.com/kubernetes/kubernetes/pull/90987 is merge d

File: staging/publishing/rules.yaml [hunk 2]
```diff
@@ -1080,189 +1080,193 @@ rules:
     - repository: client-go
       branch: release-1.18
 
-- destination: kube-controller-manager
+- destination: cloud-provider
```
Comment (andrewsykim) [inline staging/publishing/rules.yaml:1083]: This is an unfortunate diff, but the verify script was complaining that cloud-provider was reference after kube-controller-manager since kube-controller-manager now depends on cloud-provider

**Comment by andrewsykim** [comment]
Ready for review

/assign @cheftako @cici37 @deads2k @liggitt

File: staging/src/k8s.io/kube-controller-manager/go.mod [hunk 3]
```diff
@@ -15,6 +16,7 @@ replace (
 	k8s.io/api => ../api
 	k8s.io/apimachinery => ../apimachinery
 	k8s.io/client-go => ../client-go
+	k8s.io/cloud-provider => ../cloud-provider
```
Comment (liggitt) [inline staging/src/k8s.io/kube-controller-manager/go.mod:19]: @cheftako did you expect this direction or the reverse? This seems related to https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/kubernetes-sig-api-machinery/hp3uQzAIOsU/njQXHAdvAAAJ ... a sketch of the expected final shape/location for config and controller code might be helpful

File: api/api-rules/violation_exceptions.list [hunk 4]
```diff
@@ -588,7 +588,6 @@ API rule violation: names_match,k8s.io/kube-controller-manager/config/v1alpha1,R
 API rule violation: names_match,k8s.io/kube-controller-manager/config/v1alpha1,SAControllerConfiguration,ConcurrentSATokenSyncs
 API rule violation: names_match,k8s.io/kube-controller-manager/config/v1alpha1,SAControllerConfiguration,RootCAFile
 API rule violation: names_match,k8s.io/kube-controller-manager/config/v1alpha1,SAControllerConfiguration,ServiceAccountKeyFile
-API rule violation: names_match,k8s.io/kube-controller-manager/config/v1alpha1,ServiceControllerConfiguration,ConcurrentServiceSyncs
```
Comment (liggitt) [inline api/api-rules/violation_exceptions.list:591]: unless this got fixed while moving, I expected this to still be flagged in the new file location

File: cmd/cloud-controller-manager/app/apis/config/v1alpha1/types.go [hunk 5]
```diff
@@ -33,7 +34,7 @@ type CloudControllerManagerConfiguration struct {
 	KubeCloudShared kubectrlmgrconfigv1alpha1.KubeCloudSharedConfiguration
```
Comment (liggitt) [inline cmd/cloud-controller-manager/app/apis/config/v1alpha1/types.go:34]: I was surprised KubeCloudSharedConfiguration didn't have to move

**Comment by liggitt** [comment]
Moving the controller loops seems reasonable. Moving the v1alpha1 config struct which is tightly tied to the kube-controller-manager v1alpha1 config is less clear to me. Nothing inside the controller code references that config, there's a single option for the number of sync workers that is passed to the controller's constructor.

Additionally, it doesn't look like all the config type linting got wired up correctly to the new location. Can we limit this PR to moving the controller loops?

**Comment by andrewsykim** [comment]
> Additionally, it doesn't look like all the config type linting got wired up correctly to the new location. Can we limit this PR to moving the controller loops?

sgtm, I'll update shortly

File: staging/publishing/rules.yaml [hunk 6]
```diff
@@ -1281,6 +1285,10 @@ rules:
       branch: master
     - repository: cloud-provider
       branch: master
+    - repository: component-base
+      branch: master
+    - repository: kube-controller-manager
```
Comment (andrewsykim) [inline staging/publishing/rules.yaml:1290]: Btw, this was one of the reasons I initially moved the versioned service controller config to `k8s.io/cloud-provider`. Now anything that imports the cloud-provider has to import kube-controller-manager. I'll have a follow-up PR to fix this with the validation/generator wired in correctly.

**Comment by andrewsykim** [comment]
@liggitt moved versioned service controller config back to kube-controller-manager.

/retest

File: staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go [hunk 7]
```diff
@@ -207,3 +207,26 @@ func shutdownInCloudProvider(ctx context.Context, cloud cloudprovider.Interface,
 
 	return shutdown, err
 }
+
+// ensureNodeExistsByProviderID checks if the instance exists by the provider id,
+// If provider id in spec is empty it calls instanceId with node name to get provider id
+func ensureNodeExistsByProviderID(ctx context.Context, instances cloudprovider.Instances, node *v1.Node) (bool, error) {
```
Comment (cici37) [inline staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:213]: I am not sure why we cannot use the same function in node_controller.go and when/where this func will be called?
If this func has to be added here, should the test also be added in node_lifecycle_controller_test.go?

**Comment by cici37** [comment]
Should `cmd/cloud-controller-manager/.import-restrictions` also be modified due to cloud provider controller movement?

File: cmd/cloud-controller-manager/app/BUILD [hunk 8]
```diff
@@ -24,6 +21,10 @@ go_library(
         "//staging/src/k8s.io/client-go/tools/leaderelection:go_default_library",
         "//staging/src/k8s.io/client-go/tools/leaderelection/resourcelock:go_default_library",
         "//staging/src/k8s.io/cloud-provider:go_default_library",
+        "//staging/src/k8s.io/cloud-provider/controllers/node:go_default_library",
+        "//staging/src/k8s.io/cloud-provider/controllers/nodelifecycle:go_default_library",
```
Comment (cheftako) [inline cmd/cloud-controller-manager/app/BUILD:25]: Little worried that the name nodelifecycle here might be confused with https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/nodelifecycle. They should not be confused with each other. cloudnodelifecycle seems cumbersome but something to prevent us crossing the streams would seem prudent.

File: staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go (hunk 7 above)
Comment (cheftako) [reply staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:213]: +1, seems like this would be a good start for a controller utils package under cloud-provider.

File: staging/publishing/rules.yaml (hunk 6 above)
Comment (cheftako) [reply staging/publishing/rules.yaml:1291]: I get that KCM has to pull in the cloud provider controllers but we should be able to break that when the extraction effort finishes. Having to pull in the KCM because you pulled in the cloud-provider is definitely wrong. That's not something we want to do with the KAS or the Kubelet. If we can't fix it with your wiring change, then I think we would be better of moving them to the new controller-manager repo for now. I agree that these controllers are generic cloud-provider controllers but ...

File: staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go (hunk 7 above)
Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:213]: A little duplication is acceptable here I think. I can't reuse ensureNodeExistsByProviderID from node_controller.go because it is in a separate package and ensureNodeExistsByProviderID is not widely used enough to justify a separate utils package for it either.

File: staging/publishing/rules.yaml (hunk 6 above)
Comment (andrewsykim) [reply staging/publishing/rules.yaml:1291]: We wouldn't have to rely on `kube-controller-manager` once the controller manager config types move to `k8s.io/controller-manager`. That seems like the proper fix here as opposed to moving all the controllers to controller-manager.

Comment (andrewsykim) [reply staging/publishing/rules.yaml:1291]: I see the PR for controller-manager got moved, I can take a stab at moving those here.

Comment (andrewsykim) [reply staging/publishing/rules.yaml:1291]: Or I think @cici37 had a PR to move them. We should either block this PR on that or merge this as is and fix it once that PR is merged.

**Review by andrewsykim — COMMENTED** [review]
> Should cmd/cloud-controller-manager/.import-restrictions also be modified due to cloud provider controller movement?

Yes, there are actually a lot of stale imports there at this point, will update in a follow-up PR.

File: cmd/cloud-controller-manager/app/BUILD (hunk 8 above)
Comment (cheftako) [reply cmd/cloud-controller-manager/app/BUILD:25]: andrewsykim points out that while the above is true, all the actual usage of this controller is done through an alias of CloudNodeLifecycle to disambiguate.

**Comment by cheftako** [comment]
Chatted with Dims and Andrew. We agree that having KCM depend on the staging/cloud-provider repo for config is not ideal.  There is an upcoming PR moving GenericControllerManagerConfiguration , KubeCloudSharedConfiguration , ServiceControllerConfiguration , etc to staging/controller-manager. Which means the external conversion gen for service controller  v1alpha1 would point to k8s.io/controller-manager instead of kube-controller-manager here https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/service/config/v1alpha1/doc.go#L19. Additionally there relevant controllers do depend on a cloud provider implementation. This will be true until cloud provider extraction completes. It does indicate that we will want to deprecate the configuration for service and route from the KCM configuration.

**Comment by andrewsykim** [comment]
@liggitt friendly ping :)

File: cmd/cloud-controller-manager/app/apis/config/v1alpha1/doc.go [hunk 9]
```diff
@@ -23,7 +23,7 @@ limitations under the License.
 // +k8s:conversion-gen=k8s.io/kubernetes/cmd/cloud-controller-manager/app/apis/config
 // +k8s:conversion-gen=k8s.io/component-base/config/v1alpha1
 // +k8s:conversion-gen=k8s.io/kubernetes/pkg/controller/apis/config/v1alpha1
-// +k8s:conversion-gen=k8s.io/kubernetes/pkg/controller/service/config/v1alpha1
+// +k8s:conversion-gen=k8s.io/cloud-provider/controllers/service/config/v1alpha1
```
Comment (liggitt) [inline cmd/cloud-controller-manager/app/apis/config/v1alpha1/doc.go:26]: I'm confused... based on https://github.com/kubernetes/kubernetes/pull/90976#issuecomment-631633404 I thought we were not moving config struct locations

File: pkg/controller/apis/config/types.go [hunk 10]
```diff
@@ -34,7 +35,6 @@ import (
 	replicasetconfig "k8s.io/kubernetes/pkg/controller/replicaset/config"
 	replicationconfig "k8s.io/kubernetes/pkg/controller/replication/config"
 	resourcequotaconfig "k8s.io/kubernetes/pkg/controller/resourcequota/config"
-	serviceconfig "k8s.io/kubernetes/pkg/controller/service/config"
```
Comment (liggitt) [inline pkg/controller/apis/config/types.go:37]: same question about config type location

**Comment by liggitt** [comment]
controller code move looks fine. I'm still not sure I understand what it means to move a v1alpha1 config type to the cloud controller module when there is no v1alpha1 top-level version present in that module. This would be a lot clearer to me if it just moved the controller code and kept the config types where they were.

File: cmd/cloud-controller-manager/app/apis/config/v1alpha1/doc.go (hunk 9 above)
Comment (andrewsykim) [reply cmd/cloud-controller-manager/app/apis/config/v1alpha1/doc.go:26]: The v1alpha1 types live in staging/src/k8s.io/kube-controller-manager/config/v1alpha1 (this is the case today as well). This just moves the generated code that was in pkg/controller/service/config/v1alpha1 to staging along with the controller. That seemed fine to me since none of the types are there (maybe this is still wrong though?) 

Before this comment https://github.com/kubernetes/kubernetes/pull/90976#issuecomment-631633404 I moved the config struct from `staging/src/k8s.io/kube-controller-manager` into `k8s.io/cloud-provider/controllers/service/config/v1alpha1`). I thought your comment was specifically for that so I reverted and left the versioned types in kube-controller-manager.

**Comment by andrewsykim** [comment]
>  I'm still not sure I understand what it means to move a v1alpha1 config type to the cloud controller module when there is no v1alpha1 top-level version present in that module. 

The v1alpha1 types have not moved, they are still in `k8s.io/kube-controller-manager/config/v1alpha1`. This PR only moves the internal config types and the generated code. Do you still prefer that the internal type and generated code stay where they are in `pkg/controller/serivce`?

File: cmd/cloud-controller-manager/app/apis/config/v1alpha1/doc.go (hunk 9 above)
Comment (liggitt) [reply cmd/cloud-controller-manager/app/apis/config/v1alpha1/doc.go:26]: > This just moves the generated code that was in pkg/controller/service/config/v1alpha1 to staging along with the controller.

The issue is that places v1alpha1 config-related code in a place that has no concept of a v1alpha1 config... that seems problematic.

**Comment by liggitt** [comment]
> Do you still prefer that the internal type and generated code stay where they are in `pkg/controller/serivce`?

I think keeping the config where it is being consumed stays more coherent for now.

**Comment by andrewsykim** [comment]
@liggitt updated to leave the config types where they are, sorry for the confusion earlier.
